### PR TITLE
Gp 4292 add metrics back

### PR DESCRIPTION
**Pull request by Christopher-Ho**
JIRA Ticket: https://jira.oicr.on.ca/browse/GP-4292

- [ X ] Updates release notes
- [ ] Updates developer documentation

Need to check equality using JsonNode objects for illumina metrics instead of string comparison because json objects could be in different order.

### CODE REVIEW

File: scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java [hunk 1]
```diff
@@ -29,7 +31,24 @@ protected final void checkDirectory(String root)
...
+
+      // For Illumina processors only
+      if (clazz.equals(IlluminaNotificationDto.class)) {
+        // Load the metrics into a Jackson JsonNode to test for equality
+        ObjectMapper MapperTest = new ObjectMapper();
+        JsonNode jsonNodeResult = MapperTest.readTree(result.getMetrics());
+        JsonNode jsonNodeReference = MapperTest.readTree(reference.getMetrics());
+        if (jsonNodeResult.equals(jsonNodeReference)) {
+          // If Json objects are equal, set string representation to also be equal
+          reference.setMetrics(result.getMetrics());
```
Comment (avarsava) [inline scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java:48]: rather than manipulate the result, fail the test early if `!jsonNodeResult.equals(jsonNodeReference)` then `result.setMetrics(null)` in all remaining cases (ie, after the if block entirely)

File: dto/src/main/java/ca/on/oicr/gsi/runscanner/dto/NotificationDto.java [hunk 2]
```diff
@@ -271,7 +271,7 @@ public boolean equals(Object obj) {
         && Objects.equals(this.containerModel, other.containerModel)
         && Objects.equals(this.healthType, other.healthType)
         && Objects.equals(this.laneCount, other.laneCount)
-        && Objects.equals(this.metrics, other.metrics)
+        // && Objects.equals(this.metrics, other.metrics) // Check equality using JsonNode instead
```
Comment (avarsava) [inline dto/src/main/java/ca/on/oicr/gsi/runscanner/dto/NotificationDto.java:274]: Remove this comment?

**Review by avarsava — APPROVED** [review]

**Review by wuall826 — APPROVED** [review]

File: dto/src/main/java/ca/on/oicr/gsi/runscanner/dto/NotificationDto.java [hunk 3]
```diff
@@ -271,7 +271,6 @@ public boolean equals(Object obj) {
         && Objects.equals(this.containerModel, other.containerModel)
         && Objects.equals(this.healthType, other.healthType)
         && Objects.equals(this.laneCount, other.laneCount)
-        && Objects.equals(this.metrics, other.metrics)
```
Comment (djcooke) [inline dto/src/main/java/ca/on/oicr/gsi/runscanner/dto/NotificationDto.java:274]: If this is removed from `equals`, it should be removed from `hashCode` too, but why remove it from `equals`?

File: scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java [hunk 4]
```diff
@@ -29,7 +31,24 @@ protected final void checkDirectory(String root)
...
+        if (!jsonNodeReference.equals(jsonNodeResult)) {
+          assertEquals(reference, result);
+        }
+        // We know metrics equal, set reference to null to match result metrics
+        reference.setMetrics(null);
+      }
+
+      // For non-Illumina processors, we delete these because changes in metrics are non-critical.
+      // For Illumina processors, we've already checked equality of metrics using JsonNode equals
+      result.setMetrics(null);
```
Comment (djcooke) [inline scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java:50]: Is this necessary? I don't think non-Illumina processors have any metrics

File: scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java [hunk 5]
```diff
@@ -29,7 +31,24 @@ protected final void checkDirectory(String root)
...
+      if (clazz.equals(IlluminaNotificationDto.class)) {
+        // Load the metrics into a Jackson JsonNode to test for equality
+        ObjectMapper MapperTest = new ObjectMapper();
+        JsonNode jsonNodeResult = MapperTest.readTree(result.getMetrics());
+        JsonNode jsonNodeReference = MapperTest.readTree(reference.getMetrics());
+        if (!jsonNodeReference.equals(jsonNodeResult)) {
+          assertEquals(reference, result);
+        }
+        // We know metrics equal, set reference to null to match result metrics
+        reference.setMetrics(null);
```
Comment (djcooke) [inline scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java:45]: This is currently not really doing anything since you removed metrics comparison from `NotificationDto#equals`. Does the String version of metrics match? If yes, don't remove from either (`reference` or `result`). If no, remove from both.

File: scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java [hunk 6]
```diff
@@ -29,7 +31,24 @@ protected final void checkDirectory(String root)
...
+
+      // For Illumina processors only
+      if (clazz.equals(IlluminaNotificationDto.class)) {
+        // Load the metrics into a Jackson JsonNode to test for equality
+        ObjectMapper MapperTest = new ObjectMapper();
+        JsonNode jsonNodeResult = MapperTest.readTree(result.getMetrics());
+        JsonNode jsonNodeReference = MapperTest.readTree(reference.getMetrics());
+        if (!jsonNodeReference.equals(jsonNodeResult)) {
+          assertEquals(reference, result);
+        }
```
Comment (djcooke) [inline scanner/src/test/java/ca/on/oicr/gsi/runscanner/scanner/processor/AbstractProcessorTest.java:43]: You're already checking `reference` vs. `result` below. You can just `assertEquals(jsonNodeReference, jsonNodeResult)` here

File: scanner/src/test/resources/illumina/130417_M00146_0021_000000000-A2WDF/reference.json [hunk 7]
```diff
@@ -1,29 +1,28 @@
...
-  "imgCycle": 242,
-  "indexLengths": [],
-  "laneCount": 1,
-  "numCycles": 242,
-  "numReads": 2,
-  "pairedEndRun": true,
-  "platform": "Illumina",
-  "poolNames": {
-    "1": "Unique_Sample_ID"
-  },
```
Comment (djcooke) [inline scanner/src/test/resources/illumina/130417_M00146_0021_000000000-A2WDF/reference.json:18]: Why was this removed?

**Review by wuall826 — APPROVED** [review]

File: scanner/src/test/resources/illumina/130417_M00146_0021_000000000-A2WDF/reference.json (hunk 7 above)
Comment (djcooke) [reply scanner/src/test/resources/illumina/130417_M00146_0021_000000000-A2WDF/reference.json:18]: Ah, we actually stopped the Illumina processor from setting these a while back because the names it was using were not correct. The tests never actually checked this value anyway though

**Review by djcooke — COMMENTED** [review]
Please add `sequencerPosition` and `sequencingKit` to `NotificationDto#equals` and `#hashcode` as they are currently missing

File: runscanner-illumina/main.cpp [hunk 8]
```diff
@@ -165,6 +166,22 @@ void plot_by_lane_wrapper(
                                                skip_empty);
 }
 
+/**
+* We want to format metric numbers which are strings to include commas to have a consistent output
+* If we imbue with "" locale - this uses the system default but doesn't guarantee we get a comma
+* We want to force a comma, regardless of system default locale settings
+*/
+template<typename CharT>
+struct Sep : public std::numpunct<CharT>
```
Comment (djcooke) [inline runscanner-illumina/main.cpp:175]: This could use a more descriptive name

File: RELEASE_NOTES.md [hunk 9]
```diff
@@ -2,7 +2,9 @@
 
 Changes:
 
+  * Added back metrics for Illumina Tests
   * Added test data for NovaSeq X
+  * Update tests to include sequencing kits
```
Comment (djcooke) [inline RELEASE_NOTES.md:7]: Add a note about the Illumina metrics number format change. This will make no difference to some, including us. Others may currently have a different format, and may be depending on that.

**Review by djcooke — APPROVED** [review]
